### PR TITLE
Fit Script Generator - Create Basic Fit Options Browser

### LINKED ISSUE #30918: Fit Script Generator - Create a Basic Fit Options Browser
### Description
The `Fit Options Browser` in the `Fit Script Generator` is overly complex by displaying a large number of fit options. We only need a couple of these options such as the `Fitting Mode`, `Max Iterations`, `Minimizer` and `Cost Function`.

This issue is to implement a `BasicFitOptionsBrowser` with only these options. I will also add tests for this new browser so its stability will hopefully be an improvement over the previously used browser.

### PR DESCRIPTION
**Pull request by robertapplin**
**Description of work.**
This PR creates a Basic Fit Options Browser which provides options for only the most basic fitting options such as `Fitting Mode`, `Max Iterations`, `Minimizer`, `Cost Function` and `Evaluation Type`. This Basic Fit Options Browser will be used in the Fit Script Generator interface.

It also adds a test for this new browser.

**To test:**
1. Comment out this line
https://github.com/mantidproject/mantid/blob/44daa7417226a56f8d1896ea6b2f2b1bc576a5d1/scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_view.py#L52
2. Go to Muon Analysis interface
3. click `Load Current Run`
4. Go to Fitting tab and click `Fit Generator`
5. Click `Add Workspaces`. Select the first workspace and click add.
6. Play with the options in the bottom right. There should be no errors.
image

7. Switching between Simultaneous and Sequential Fitting Mode should show and hide a column in the FunctionTree, and a function index in the data table.

*No release notes required as this browser is not yet accessible to users*

Fixes #30918

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/widgets/common/src/BasicFitOptionsBrowser.cpp [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+std::string BasicFitOptionsBrowser::getIntProperty(QtProperty *prop) const {
+  return QString::number(m_intManager->value(prop)).toStdString();
+}
+
+void BasicFitOptionsBrowser::setStringEnumProperty(QtProperty *prop,
+                                                   std::string const &value) {
+  auto const i =
+      m_enumManager->enumNames(prop).indexOf(QString::fromStdString(value));
+  if (i >= 0)
```
Comment (DannyHindson) [inline qt/widgets/common/src/BasicFitOptionsBrowser.cpp:198]: Is it worth throwing an invalid argument exception like you do lower down if the property isn't found? Not 100% sure but maybe if someone added a new property and forgot to do m_enumManager->addProperty(XXX) this exception might be useful

**Review by DannyHindson — CHANGES_REQUESTED** [review]
I've tested this and all works fine. Just want to double check on the scope of this, it seems like you've not implemented the "Generate Fit Script" button yet - is that correct?

I've added a few detail level comments plus I've got one general question. Richard did something vaguely similar to this in the engineering diffraction UI not too long ago (Interfaces, Diffraction, Engineering Diffractoin and then Fitting tab). He added a new method FitPropertyBrowser::removePropertiesFromSettingsBrowser to achieve his cut down list of properties rather than creating a new XXXFitOptionsBrowser class. Is that solution not applicable here because you're not using the FitPropertyBrowser?

File: qt/widgets/common/src/BasicFitOptionsBrowser.cpp [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+#include "MantidQtWidgets/Common/QtPropertyBrowser/qteditorfactory.h"
+#include "MantidQtWidgets/Common/QtPropertyBrowser/qtpropertymanager.h"
+#include "MantidQtWidgets/Common/QtPropertyBrowser/qttreepropertybrowser.h"
+
+#include <QVBoxLayout>
+
+namespace {
+
+using namespace Mantid::API;
```
[the remark this answers is not in the document]
Comment (robertapplin) [reply qt/widgets/common/src/BasicFitOptionsBrowser.cpp:20]: I can see that too, not sure why 🤔

**Comment by robertapplin** [comment]
> I've tested this and all works fine. Just want to double check on the scope of this, it seems like you've not implemented the "Generate Fit Script" button yet - is that correct?
> 
> I've added a few detail level comments plus I've got one general question. Richard did something vaguely similar to this in the engineering diffraction UI not too long ago (Interfaces, Diffraction, Engineering Diffractoin and then Fitting tab). He added a new method FitPropertyBrowser::removePropertiesFromSettingsBrowser to achieve his cut down list of properties rather than creating a new XXXFitOptionsBrowser class. Is that solution not applicable here because you're not using the FitPropertyBrowser?

Thanks for taking a look

Yes the "Fit Generator" button is currently hidden so users cannot access this interface until it has been completed. If you comment out line 52 of `fitting_tab_view.py` then you can open the interface.

Yes I believe the `FitProperyBrowser` and `FitOptionsBrowser` are different things. I think Roman created the `FitOptionsBrowser` more recently as a replacement to the `FitPropertyBrowser` as its file is > 3000 lines. It would probably be possible to implement a similar method in the `FitOptionsBrowser` to remove the properties but I preferred something more concrete. At some point I would like to get the `FitOptionsBrowser` to inherit from the `BasicFitOptionsBrowser` to reuse the code and then we have two concrete classes which we can choose between whether we want all the bells and whistles or just the basics.

The `BasicFitOptionsBrowser` also hides the sub-properties within the Minimizer property to simplify things even more.

**Comment by DannyHindson** [comment]
> > I've tested this and all works fine. Just want to double check on the scope of this, it seems like you've not implemented the "Generate Fit Script" button yet - is that correct?
> > I've added a few detail level comments plus I've got one general question. Richard did something vaguely similar to this in the engineering diffraction UI not too long ago (Interfaces, Diffraction, Engineering Diffractoin and then Fitting tab). He added a new method FitPropertyBrowser::removePropertiesFromSettingsBrowser to achieve his cut down list of properties rather than creating a new XXXFitOptionsBrowser class. Is that solution not applicable here because you're not using the FitPropertyBrowser?
> 
> Thanks for taking a look
> 
> Yes the "Fit Generator" button is currently hidden so users cannot access this interface until it has been completed. If you comment out line 52 of `fitting_tab_view.py` then you can open the interface.
> 
> Yes I believe the `FitProperyBrowser` and `FitOptionsBrowser` are different things. I think Roman created the `FitOptionsBrowser` more recently as a replacement to the `FitPropertyBrowser` as its file is > 3000 lines. It would probably be possible to implement a similar method in the `FitOptionsBrowser` to remove the properties but I preferred something more concrete. At some point I would like to get the `FitOptionsBrowser` to inherit from the `BasicFitOptionsBrowser` to reuse the code and then we have two concrete classes which we can choose between whether we want all the bells and whistles or just the basics.
> 
> The `BasicFitOptionsBrowser` also hides the sub-properties within the Minimizer property to simplify things even more.

Thanks that makes sense. Yes I'd managed to follow your instructions about uncommenting the line to access the Fit Generator screen. It was just the button once you'd got inside the Fit Generator screen that I was asking about. Sounds like this screen is deliberately not finished yet though so that's all good.

It does seem like some refactoring to avoid code duplication and to enable more customisation would be helpful at some point. The inheritance you mention sounds good plus there are instrument specific versions of parts of the fit property browser for Muon, indirect and engineering diffraction where it feels like one solution that supported some overrides would be better than three parallel approaches. But that's for another time. Thanks for clarifying.

**Comment by robertapplin** [comment]
Ahh ok my apologies for that misunderstanding @DannyHindson , I forgot that button even existed! You are right that button isn't implemented yet and the interface is deliberately unfinished. Thanks for taking a look at that 👍

**Review by DannyHindson — APPROVED** [review]
Thanks for making the changes. This looks good now

**Comment by martyngigg** [comment]
I agree with Danny here in that we need to watch out for many duplications of the same thing. I think at this point we are probably in a better position to understand what we want from this "fit options" type widget across all of the uses? We need to be slightly careful in overusing inheritance just for code reuse as this can make future changes more difficult when child classes auomatically inherit behaviour (https://en.wikipedia.org/wiki/Composition_over_inheritance). It sounds a bit like we maybe need better abilities to configure such a widget from the outside without having to necessarily generate new classes?

Can I suggest we generate a task to consider how we amalagate these things and replace the `FitPropertyBrowser` with something more maintainable going forward? At this point I think it would be an investigate and generate plan/design task.

**Comment by robertapplin** [comment]
I think you are both right. In retrospect this probably wasn't the best solution for adding flexibility to the fit options. I have created an issue to fix this #30992

**Pull request merged**